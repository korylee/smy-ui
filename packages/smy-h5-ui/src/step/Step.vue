<template>
  <div :class="bem([status])" @click="onClick">
    <div :class="bem('header')">
      <div :class="bem('header-line')"></div>
      <div :class="bem('header-icon', { dot })">
        <slot name="icon">{{ dot ? '' : currentIndex + 1 }}</slot>
      </div>
    </div>
    <div :class="bem('main')">
      <div :class="bem('main-title')">
        <slot name="title">{{ title }}</slot>
      </div>
      <div v-if="content || hasSlot()" :class="bem('main-content')">
        <slot><span v-html="content"></span></slot>
      </div>
    </div>
  </div>
</template>

<script>
import { getListener } from '../_mixins/listeners'
import { createChildrenMixin } from '../_mixins/relation'
import { createNamespace } from '../_utils/vue/create'
import { SlotsMixin } from '../_utils/vue/slots'
import { props } from './props'

const [name, bem] = createNamespace('step')

export default {
  name,
  mixins: [createChildrenMixin('steps', { children: 'step' }), SlotsMixin],
  props,
  computed: {
    dot({ steps }) {
      return steps.progressDot
    },
    currentIndex({ index, steps }) {
      const { reverse, step } = steps
      const length = step?.length ?? 0
      return reverse ? length - index - 1 : index
    },
    status({ steps, currentIndex }) {
      const { current } = steps
      const isFinish = currentIndex < +current
      if (isFinish) return 'finish'
      return currentIndex === +current ? 'process' : 'wait'
    },
    direction({ steps }) {
      return steps.direction
    },
  },
  methods: {
    bem,
    onClick() {
      const { steps, index } = this
      const clickStep = getListener.call(steps, 'click-step')
      clickStep(index)
    },
  },
}
</script>

<style lang="less">
@import '../_styles/common.less';
@import './step.less';
</style>
