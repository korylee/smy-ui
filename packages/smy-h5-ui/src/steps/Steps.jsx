import { getSlot } from '../_utils/vue/slots'
import { createParentMixin } from '../_mixins/relation'
import { props } from './props'

import './steps.less'
import { createNamespace } from '../_utils/vue/create'

const [name, bem] = createNamespace('steps')

export default {
  name,
  mixins: [createParentMixin('steps', { children: 'step' })],
  props,
  computed: {
    direction({ vertical }) {
      return vertical ? 'vertical' : 'horizontal'
    },
  },
  render() {
    const vm = this
    const { direction, progressDot: dot } = vm

    return <div class={bem({ dot, [direction]: direction })}>{getSlot(vm)}</div>
  },
}
