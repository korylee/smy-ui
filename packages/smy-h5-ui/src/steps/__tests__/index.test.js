import { mount } from '@vue/test-utils'
import Step from '../../step'
import example from '../example'
import Steps from '..'
import Vue from 'vue'

test('test steps example', () => {
  const wrapper = mount(example)

  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
})

test('test steps and step plugin', () => {
  Vue.use(Steps).use(Step)

  expect(Vue.component(Steps.name)).toBeTruthy()
  expect(Vue.component(Step.name)).toBeTruthy()
})

test('test step direction prop', async () => {
  const template = `
    <smy-steps :vertical="vertical">
      <smy-step>步骤1</smy-step>
      <smy-step>步骤2</smy-step>
    </smy-steps>
  `
  const wrapper = mount(
    {
      data: () => ({ vertical: false }),
      components: { SmySteps: Steps, SmyStep: Step },
      template,
    },
    { attachTo: document.body }
  )
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setData({ vertical: true })
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test step event', async () => {
  const clickStep = jest.fn()
  const template = `
  <smy-steps :current="current" @click-step="clickStep">
   <smy-step>步骤1</smy-step>
   <smy-step>步骤2</smy-step>
 </smy-steps>
`
  const wrapper = mount(
    {
      template,
      data: () => ({ current: 1 }),
      components: { SmySteps: Steps, SmyStep: Step },
      methods: {
        clickStep,
      },
    },
    { attachTo: document.body }
  )
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.find('.smy-step').trigger('click')
  expect(clickStep).toHaveBeenCalledTimes(1)
  wrapper.destroy()
})
