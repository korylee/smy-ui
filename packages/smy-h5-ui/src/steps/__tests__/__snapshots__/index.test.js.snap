// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test step direction prop 1`] = `
"<div class=\\"smy-steps smy-steps--horizontal\\">
  <div class=\\"smy-step smy-step--finish\\">
    <div class=\\"smy-step__header\\">
      <div class=\\"smy-step__header-line\\"></div>
      <div class=\\"smy-step__header-icon\\">
        <div class=\\"smy-step__header-icon-seq\\">0</div>
      </div>
    </div>
    <div class=\\"smy-step__main\\">
      <div class=\\"smy-step__main-title\\"></div>
      <div class=\\"smy-step__main-content\\">步骤1</div>
    </div>
  </div>
  <div class=\\"smy-step smy-step--finish\\">
    <div class=\\"smy-step__header\\">
      <div class=\\"smy-step__header-line\\"></div>
      <div class=\\"smy-step__header-icon\\">
        <div class=\\"smy-step__header-icon-seq\\">0</div>
      </div>
    </div>
    <div class=\\"smy-step__main\\">
      <div class=\\"smy-step__main-title\\"></div>
      <div class=\\"smy-step__main-content\\">步骤2</div>
    </div>
  </div>
</div>"
`;

exports[`test step direction prop 2`] = `
"<div class=\\"smy-steps smy-steps--vertical\\">
  <div class=\\"smy-step smy-step--process\\">
    <div class=\\"smy-step__header\\">
      <div class=\\"smy-step__header-line\\"></div>
      <div class=\\"smy-step__header-icon\\">
        <div class=\\"smy-step__header-icon-seq\\">1</div>
      </div>
    </div>
    <div class=\\"smy-step__main\\">
      <div class=\\"smy-step__main-title\\"></div>
      <div class=\\"smy-step__main-content\\">步骤1</div>
    </div>
  </div>
  <div class=\\"smy-step smy-step--wait\\">
    <div class=\\"smy-step__header\\">
      <div class=\\"smy-step__header-line\\"></div>
      <div class=\\"smy-step__header-icon\\">
        <div class=\\"smy-step__header-icon-seq\\">2</div>
      </div>
    </div>
    <div class=\\"smy-step__main\\">
      <div class=\\"smy-step__main-title\\"></div>
      <div class=\\"smy-step__main-content\\">步骤2</div>
    </div>
  </div>
</div>"
`;

exports[`test step event 1`] = `
"<div class=\\"smy-steps smy-steps--horizontal\\">
  <div class=\\"smy-step smy-step--finish\\">
    <div class=\\"smy-step__header\\">
      <div class=\\"smy-step__header-line\\"></div>
      <div class=\\"smy-step__header-icon\\">
        <div class=\\"smy-step__header-icon-seq\\">0</div>
      </div>
    </div>
    <div class=\\"smy-step__main\\">
      <div class=\\"smy-step__main-title\\"></div>
      <div class=\\"smy-step__main-content\\">步骤1</div>
    </div>
  </div>
  <div class=\\"smy-step smy-step--finish\\">
    <div class=\\"smy-step__header\\">
      <div class=\\"smy-step__header-line\\"></div>
      <div class=\\"smy-step__header-icon\\">
        <div class=\\"smy-step__header-icon-seq\\">0</div>
      </div>
    </div>
    <div class=\\"smy-step__main\\">
      <div class=\\"smy-step__main-title\\"></div>
      <div class=\\"smy-step__main-content\\">步骤2</div>
    </div>
  </div>
</div>"
`;

exports[`test steps example 1`] = `
"<div class=\\"steps-example\\">
  <div class=\\"app-demo-title\\">基础使用</div>
  <div class=\\"steps-wrapper smy-steps smy-steps--horizontal\\">
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon\\">
          <div class=\\"smy-step__header-icon-seq\\">0</div>
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\">已完成</div>
        <div class=\\"smy-step__main-content\\"><span>这里是该步骤的描述信息加长加长加长加长加长加长加长</span></div>
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon\\">
          <div class=\\"smy-step__header-icon-seq\\">0</div>
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <!---->
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon\\">
          <div class=\\"smy-step__header-icon-seq\\">0</div>
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <!---->
      </div>
    </div>
  </div>
  <div class=\\"steps-button\\"><button>下一步</button></div>
  <div class=\\"app-demo-title\\">基础使用(点状)</div>
  <div class=\\"steps-wrapper smy-steps smy-steps--dot smy-steps--horizontal\\">
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <!---->
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <!---->
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <!---->
      </div>
    </div>
  </div>
  <div class=\\"steps-button\\"><button>下一步</button></div>
  <div class=\\"app-demo-title\\">竖向</div>
  <div class=\\"steps-wrapper smy-steps smy-steps--vertical\\">
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon\\">
          <div class=\\"smy-step__header-icon-seq\\">0</div>
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\">已完成</div>
        <div class=\\"smy-step__main-content\\"><span>这里是该步骤的描述信息加长加长加长加长加长加长加长</span></div>
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon\\">
          <div class=\\"smy-step__header-icon-seq\\">0</div>
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <!---->
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon\\">
          <div class=\\"smy-step__header-icon-seq\\">0</div>
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <!---->
      </div>
    </div>
  </div>
  <div class=\\"steps-button\\"><button>下一步</button></div>
  <div class=\\"app-demo-title\\">点状</div>
  <div class=\\"steps-wrapper smy-steps smy-steps--dot smy-steps--vertical\\">
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\">已完成</div>
        <div class=\\"smy-step__main-content\\"><span>这里是该步骤的描述信息加长加长加长加长加长加长加长</span></div>
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <div class=\\"smy-step__main-content\\"><span>fasdfad</span></div>
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <div class=\\"smy-step__main-content\\"><span>fasdfad</span></div>
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <div class=\\"smy-step__main-content\\"><span>fasdfad</span></div>
      </div>
    </div>
  </div>
  <div class=\\"app-demo-title\\">翻转</div>
  <div class=\\"steps-wrapper smy-steps smy-steps--dot smy-steps--vertical\\">
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"><span>收货地址：广东省深圳市福田区梅林街道新一代产业园5栋</span></div>
        <div class=\\"smy-step__main-content\\">
          刘一飞 18889898888
        </div>
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <div class=\\"smy-step__main-content\\"><span>fasdfad</span></div>
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <div class=\\"smy-step__main-content\\"><span>fasdfad</span></div>
      </div>
    </div>
    <div class=\\"smy-step smy-step--finish\\">
      <div class=\\"smy-step__header\\">
        <div class=\\"smy-step__header-line\\"></div>
        <div class=\\"smy-step__header-icon smy-step__header-icon--dot\\">
          <!---->
        </div>
      </div>
      <div class=\\"smy-step__main\\">
        <div class=\\"smy-step__main-title\\"></div>
        <div class=\\"smy-step__main-content\\"><span>fasdfad</span></div>
      </div>
    </div>
  </div>
</div>"
`;
